Added unittesting framework

shuhaowu opened the pull request:
The requirements.txt has changed, so please updated accordingly..

This is still work in progress. So do not merge until I mark it as ready.

shuhaowu committed:
Added unittesting framework

The requirements.txt has changed, so please updated accordingly..

travellingprog commented:
I haven't reviewed everything yet, and tbh it'll take me a bit of time to understand the codebase, because I'm not familiar with Flask.

For now, I'll say that your change in requirements.txt seems fine, since your fork of Flask-login essentially just adds two test methods. Normally I'd say it's not wise to your own personal fork as a requirement, but it doesn't look like the original flask-login is going to get updated anytime soon. 

I don't know how you'll make use of those test methods, since they don't bear the same name as the methods you want to emulate... but I'm guessing that there's something in unittest that allows you to re-direct a method call?

shuhaowu committed:
Initial API testing

This is also a pretty massive change. There is another requirements.txt
change as I found a bug in a library that we depend on.

shuhaowu commented:
I made some more changes. I'll hope to comment some more of this later, but you can roughly see how I'm using these now. 

I also left some comment inline (on github). You should probably take a look.

shuhaowu committed:
Another round of changes

I forecast some more big changes coming for the test utils and how we
test things. Should settle down soon, though

Found some more weird bugs in leveldbkit so we need to update
requirements.txt again.

shuhaowu committed:
Feed API tests are mostly complete.

shuhaowu committed:
Added profile tests

shuhaowu committed:
Changed test cast layout a little

shuhaowu committed:
Added test to cover most of the todo_api

shuhaowu commented:
I feel like this is stable enough to merge. So I'm gonna just merge this. There will be some requirements.txt updates but that shouldn't be too big of an issue.

I know we're gonna depend on some of my patches, but I hope to land some of them in their respective master if that is possible.

This is so that we can move forward and work on new features. I've been spending the last 2 weekends working this and nothing else essentially.
